synapse/storage/state.py

```diff
@@ -356,6 +375,166 @@ def get_member_split(self) -> Tuple["StateFilter", "StateFilter"]:
+        written as E(F), then the resultant state filter bears this property:
+
+            E(self) ∖ E(subtrahend)
+                ⊆ E(approx_difference(self, subtrahend))
+                ⊆ E(self)
+
+
+        This implementation attempts to return the narrowest such state filter.
+        In the case that `self` contains wildcards for state types where
+        `subtrahend` contains specific state keys, an approximation must be made:
+        the resultant state filter keeps the wildcard, as state filters are not
+        able to express 'all state keys except some given examples'.
+        e.g.
+            StateFilter(m.room.member -> None (wildcard))
+                minus
+            StateFilter(m.room.member -> {'@wombat:example.org'})
+                is approximated as
+            StateFilter(m.room.member -> None (wildcard))
+                which satisfies the condition that the resultant state filter
+                is an over-approximation.
+        """
+
+        # We first transform self and subtrahend into an alternative representation:
+        #   - whether or not they include all events to begin with ('all')
+        #   - if so, which event types are excluded? ('excludes')
+        #   - which entire event types to include ('wildcards')
+        #   - which concrete state keys to include ('concrete state keys')
+        (self_all, self_excludes), (
+            self_wildcards,
+            self_concrete_keys,
+        ) = self.decompose_into_four_parts()
+        (sub_all, sub_excludes), (
+            sub_wildcards,
+            sub_concrete_keys,
+        ) = subtrahend.decompose_into_four_parts()
+
+        # Start with an estimate of the difference based on self
+        new_all = self_all
+        # Wildcards from the subtrahend can be added to the exclusion filter
+        new_excludes = self_excludes | sub_wildcards
+        # We remove wildcards that appeared as wildcards in the subtrahend
+        new_wildcards = self_wildcards - sub_wildcards
+        # We filter out the concrete state keys that appear in the subtrahend
+        # as wildcards or concrete state keys.
+        new_concrete_keys = {
+            (state_type, state_key)
+            for (state_type, state_key) in self_concrete_keys
+            if state_type not in sub_wildcards
+        } - sub_concrete_keys
+
+        if self_all:
+            self_concretes_or_wild_types = self_wildcards.union(
+                typ for typ, _key in self_concrete_keys
+            )
+            # If self starts with all, then we add as wildcards any
+            # types which appear in the subtrahend's exclusion filter but do
+            # not already appear in self's concrete or wild types.
+            # This is needed to handle the fact that `include_others` includes
+            # all unspecified state types.
+            new_wildcards |= sub_excludes.difference(self_concretes_or_wild_types)
```

Comment:
```suggestion
            # If self starts with all, then we add as wildcards any
            # types which appear in the subtrahend's exclusion filter (but
            # aren't in the self exclusion filter). This is as the subtrahend
            # filter will return everything BUT the types in its exclusion, so
            # we need to add those excluded types that also match the self
            # filter as wildcard types in the new filter.
            new_wildcards |= sub_excludes.difference(self_exclusion)
```

(Note that `self_exclusion` is the same as the union of wildcard types and concrete types in our case, due to the way `include_others` works)

I'd be also tempted to move this into the `if sub_all` section, as it really only makes sense if both sides are `ALL`, though it doesn't functionally matter as if `sub_all` is false then `sub_excludes` is empty.